Update @wordpress JavaScript packages for WordPress 6.0 Beta1 release

adamziel opened the pull request:
## What?
This huge PR is required to ship WordPress 6.0 Beta 1. It is part of the process described in the [developer handbook](https://make.wordpress.org/core/handbook/about/release-cycle/block-editor-release-process-for-major-releases/).

## Why?
The WordPress repository contains the Gutenberg files, some of them in a pre-built form.

Releasing a new WordPress version means bringing a new version of the Gutenberg code from the relevant Gutenberg release. In this case, it is [v13.0.0-rc.3](https://github.com/WordPress/gutenberg/releases/tag/v13.0.0-rc.3).

It may seem somewhat awkward to ship a Gutenberg Release Candidate instead of the stable version – it is intentional, though. That's how the process is lined up in the [release calendar](https://make.wordpress.org/core/6-0/). Beta 1 is scheduled for today, April 12th, and v13.0.0-rc.3 is the latest available Gutenberg release. The stable Gutenberg version v13.0.0 will only be released later this week and included in an upcoming WordPress 6.0 Beta 2 on April 19th.

## How?

**Step 0: Pre-requisites**

There was some grunt work to perform before this PR could be even created:

* [Include relevant blocks in the Gutenberg release](https://github.com/WordPress/gutenberg/pull/40186)
* Release Gutenberg [v13.0.0-rc.3](https://github.com/WordPress/gutenberg/releases/tag/v13.0.0-rc.3).
* [Backport the PHP changes from the Gutenberg plugin into wordpress-develop](https://github.com/WordPress/gutenberg/issues/39889)

Every next step depends upon these.

**Step 1: Update WordPress packages**

I've updated all the relevant packages
```
$ npm install
$ npm run wp-packages-update
```

**Step 2: Update blocks list** 

I've made adjustments in all the files that include a list of blocks in some form:

- tools/webpack/blocks.js
- wp-content/plugin/lib/blocks.php
- wp-includes/blocks/index.php

Then also unregistered the init hooks in `tests/phpunit/includes/functions.php`

**Step 3: build**

I've ran the production and the development build:

```
# Run WordPress production build in the repo root
$ cd wordpress-develop
$ npm run build

# Run WordPress dev build in src/
$ cd src
$ npm run build:dev
```

**Step 4: Whack-a-mole**

I've spent a while iterating on any emerging problems and CI failures with @gziolo. It is partially documented in the discussion below:

* [Unit tests failed – the fixtures for server-side rendered blocks do not include the `wp-container-\d+` class](https://github.com/WordPress/wordpress-develop/pull/2564#issuecomment-1095828815)
* [Unit tests failed – same as above, but for the `wp-gallery-\d+` class](https://github.com/WordPress/wordpress-develop/pull/2564#issuecomment-1096506853)

**Step 5: SVN commit**

@gziolo listed the `svn:ignore` steps required to commit these changes to the SVN repo.

https://github.com/WordPress/wordpress-develop/pull/2564#issuecomment-1096809225

Trac ticket: https://core.trac.wordpress.org/ticket/55505 (created to keep track of the process on the WordPress SVN/Trac side of things)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamziel committed:
Update WordPress packages, backport new blocks and updates to the existing ones

adamziel committed:
Adjust webpack configuration to reflect changes from Gutenberg

* Add the @wordpress/preferences package
* Rename GUTENBERG_PHASE1 to IS_GUTENBERG_PLUGIN
* Remove COMPONENT_SYSTEM_PHASE

adamziel committed:
Regenerate package-lock.json after rebasing

gziolo commented on tools/webpack/packages.js:
```diff
@@ -149,10 +149,8 @@ module.exports = function( env = { environment: 'production', watch: false, buil
-				// Inject the `COMPONENT_SYSTEM_PHASE` global, used for controlling Component System roll-out.
-				'process.env.COMPONENT_SYSTEM_PHASE': 0,
+				// Inject the `IS_GUTENBERG_PLUGIN` global, used for feature flagging.
+				'process.env.IS_GUTENBERG_PLUGIN': 1,
```
This relfects changes applied in the Gutenberg plugin.

gziolo replied:
Actually, this should be set to `false`. We should exclude all changes from the Guetenberg plugin that are experimental.

adamziel replied:
Done in https://github.com/WordPress/wordpress-develop/pull/2564/commits/735f98785412c02ca3fd7c1204e6908264d16375

gziolo commented on package.json:
```diff
@@ -76,59 +76,60 @@
+		"@wordpress/viewport": "4.4.0",
+		"@wordpress/warning": "2.6.0",
+		"@wordpress/widgets": "2.4.0",
+		"@wordpress/wordcount": "3.6.0",
```
We should also updated `react` and `react-dom` to v17.0.2 in this file and here:

https://github.com/WordPress/wordpress-develop/blob/cab6fec683eba7bcecad119d3e9d38850255033b/src/wp-includes/script-loader.php#L106-L107

ramonjd commented:
👋 

[Failing tests](https://github.com/WordPress/wordpress-develop/runs/5976481941?check_suite_focus=true) such as these –

```bash

5) Tests_Blocks_Render::test_do_block_output with data set #22 ('core__gallery__columns.html', 'core__gallery__columns.server.html')
File '/var/www/tests/phpunit/includes/../data/blocks/fixtures/core__gallery__columns.html' does not match expected value
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 '
-<ul class="wp-block-gallery columns-1 is-cropped">
+<ul class="wp-container-12 wp-block-gallery-11 wp-block-gallery columns-1 is-cropped">
 	<li class="blocks-gallery-item">
 		<figure>
 			<img src="https://cldup.com/uuUqE_dXzy.jpg" alt="title" />

/var/www/tests/phpunit/tests/blocks/render.php:228
```
– relate to a change in the Columns block, where we added layout support. See comment: https://github.com/WordPress/gutenberg/pull/37360#issuecomment-1017009616

Blocks that opt into layout support  are gifted generated classnames 🎁 of `wp_unique_id( 'wp-container-' )`

One possible approach is to update `test_do_block_output` in [Tests_Blocks_Render](https://github.com/ramonjd/wordpress-develop/blob/40ac5de838be0d75ec4f24bf400f04516745a736/tests/phpunit/tests/blocks/render.php#L212) to swap the id for something we expect. It's how Gutenberg approaches it when [testing generated classnames in elements support.](https://github.com/WordPress/gutenberg/blob/trunk/phpunit/class-elements-test.php#L15)


```php
	public function test_do_block_output( $html_filename, $server_html_filename ) {
		$html_path        = self::$fixtures_dir . '/' . $html_filename;
		$server_html_path = self::$fixtures_dir . '/' . $server_html_filename;

		foreach ( array( $html_path, $server_html_path ) as $filename ) {
			if ( ! file_exists( $filename ) ) {
				throw new Exception( "Missing fixture file: '$filename'" );
			}
		}

		$html = do_blocks( self::strip_r( file_get_contents( $html_path ) ) );
		// If blocks opt into Gutenberg's layout implementation
		// the container will receive an added classname of `wp_unique_id( 'wp-container-' )`
		// so we need to normalize the random id.
		$normalized_html = preg_replace( '/wp-container-\d+/', 'wp-container-1', $html );
		$expected_html   = self::strip_r( file_get_contents( $server_html_path ) );

		$this->assertSame(
			$expected_html,
			$normalized_html,
			"File '$html_path' does not match expected value"
		);
	}
```

And then in the columns (and any other offending fixtures)

```html
<!-- wp:column -->
<div class="wp-container-1 wp-block-column">
	<!-- wp:paragraph -->
	<p>Column One, Paragraph One</p>
	<!-- /wp:paragraph -->
	<!-- wp:paragraph -->
	<p>Column One, Paragraph Two</p>
	<!-- /wp:paragraph -->
</div>
<!-- /wp:columns -->
```

ocean90 commented:
> And then in the columns (and any other offending fixtures)

The new `wp-container-` class is expected because the layout option for the column block was enabled in https://github.com/WordPress/gutenberg/pull/39422.

adamziel committed:
Normalize the HTML to replace wp_unique_id in certain classnames with 1

This relates to a change in the Columns block that added layout support.
See https://github.com/WordPress/gutenberg/pull/37360#issuecomment-1017009616.
Blocks that opt into layout support are gifted generated classnames of wp_unique_id( 'wp-container-' ).
One possible approach is to update test_do_block_output in Tests_Blocks_Render to swap the id for something
we expect. It's how Gutenberg approaches it when testing generated classnames in elements support.

adamziel commented:
### Problem
There's one more puzzle piece to address in regard to class names – the `wp-block-gallery-11` thing:

```
-<ul class="wp-block-gallery columns-1 is-cropped">
+<ul class="wp-container-12 wp-block-gallery-11 wp-block-gallery columns-1 is-cropped">
```

It was introduced specifically for the gallery block in https://github.com/WordPress/gutenberg/pull/38164

Fixtures from the Gutenberg trunk don't seem to acknowledge that class name: https://github.com/WordPress/gutenberg/blob/trunk/test/integration/fixtures/blocks/core__gallery__columns.html

The integration tests implemented in JavaScript pass without a problem, presumably because `wp_unique_id` is never ran in these code paths.

### Solution

* I might go for a similar preg_replace exception as for the `wp-container-` thing for now. 
* ~~Adjusting the fixtures here might require a backport to the gutenberg repo~~ server fixtures are only present in this repo.

adamziel committed:
Normalize the gallery block classname of wp_unique_id( 'wp-block-gallery-' )

Replace the random-ish output such as wp-block-gallery-12 with a
predictable one like wp-block-gallery-1

adamziel committed:
Add the newly expected wp-container-1 CSS class to column block fixtures

adamziel committed:
Add the newly expected wp-container-1 CSS class to columns block fixtures

adamziel committed:
Add the newly expected wp-container-1 CSS class to gallery block fixtures

adamziel committed:
Add the newly expected wp-iblock-gallery-1 CSS class to gallery block fixtures, and make the PHP tests mindful of the randomness

adamziel committed:
Add the missing CSS classes to the gallery columns fixture

adamziel committed:
Add the newly expected wp-container-1 CSS class to the deprecated columns fixture

adamziel committed:
Update the packages to their latest versions

adamziel committed:
Update script-loader-packages.php

adamziel committed:
Update React to 17.0.2

adamziel committed:
Set IS_GUTENBERG_PLUGIN to false

gziolo commented on tools/webpack/blocks.js:
```diff
@@ -71,13 +85,15 @@ module.exports = function( env = { environment: 'production', watch: false, buil
 		'heading',
 		'html',
 		'list',
+		'list-item',
```
We miss one more block:

Post Author Biography

https://github.com/WordPress/gutenberg/pull/38269

gziolo replied:
List Item is not used at the moment but we probably need to include it.

adamziel replied:
Good spot! Just added in https://github.com/WordPress/wordpress-develop/pull/2564/commits/5e0bbf43be14a64c82332be4cc028c6237a3b84a

adamziel committed:
Rebuild dev packages

adamziel committed:
Add the missing post-author-biography block

adamziel committed:
Add the missing blocks to blocks/index.php

gziolo commented on src/wp-includes/blocks/index.php:
```diff
@@ -7,12 +7,25 @@
 require ABSPATH . WPINC . '/blocks/gallery.php';
+require ABSPATH . WPINC . '/blocks/home-link.php';
 require ABSPATH . WPINC . '/blocks/image.php';
+require ABSPATH . WPINC . '/blocks/index.php';
```
Inception alert! ➿ 

We need to remove this one as it tries to import self 😄

adamziel committed:
Remove unnecessary self-require

adamziel committed:
remove reinitialization of block registration

adamziel committed:
Add two missing remove_action hooks

gziolo commented:
We still need to use `svn:ignore` in newly added folders with new blocks to align with the following rules in `.gitignore`:

```
/src/wp-includes/blocks/**/*.css
/src/wp-includes/blocks/**/*.js
/src/wp-includes/blocks/**/*.js.map
```

adamziel commented:
The CSS styles are not loading properly in the site editor on this branch – notice how the site logo block looks like:

<img width="1510" alt="Screenshot 2022-04-12 at 16 26 13" src="https://user-images.githubusercontent.com/205419/162987970-2a761132-2177-4663-9b09-f587f04af242.png">

The problem does not occur in trunk:

![CleanShot 2022-04-12 at 16 39 01](https://user-images.githubusercontent.com/205419/162988035-aeb477ae-bc55-487b-826d-17be1a968b9f.png)

It broke in the very beginning of this branch - right after updating the WordPress packages: c50d107ea6

gziolo commented:
Create `ignorelist.txt`:
```txt
*.js.map
*.js
*.css
```

Run in every new block folder:
```bash
svn propset svn:ignore -F ingorelist.txt .
```

adamziel commented:
The CSS issue is due to editor.css files not being loaded in the site editor. Broken on the left, working on the right:

![CleanShot 2022-04-12 at 16 56 11](https://user-images.githubusercontent.com/205419/162993635-cb5cfcdd-9622-4245-a1d9-b62772855c6b.png)
![CleanShot 2022-04-12 at 16 54 36](https://user-images.githubusercontent.com/205419/162993673-80514951-dea8-4d8f-b0a5-15df63558241.png)
![CleanShot 2022-04-12 at 16 55 15](https://user-images.githubusercontent.com/205419/162993655-14d70147-b13a-4f7a-ab94-98c1dce63015.png)

This PR doesn't change anything related to the site editor other than bumping the package version in the package.json from `3.0.27` to `4.3.1` – that must be the problem, then?

gziolo commented:
Committed with https://core.trac.wordpress.org/changeset/53157. Awesome work @adamziel wrangling all those tricky changes. Let's tackle the issues with styles in the site editor with a new PR.

adamziel commented:
The CSS issue seems to be a missing backport of https://github.com/WordPress/gutenberg/pull/37193

`wordpress-develop` trunk relies on `__editorAssets` which is exactly what the pr above replaced with  `__unstableResolvedAssets`. cc @ellatrix

gziolo commented:
> The CSS issue seems to be a missing backport of [WordPress/gutenberg#37193](https://github.com/WordPress/gutenberg/pull/37193)

Fixed with https://github.com/WordPress/wordpress-develop/commit/1d086c57d5bb1da661d675fae8acca94390a4ffa.

gziolo commented:
Commit: https://core.trac.wordpress.org/changeset/53160.
